### PR TITLE
Add pymc support to predictive explorer

### PR DESCRIPTION
**Pull request by rohanbabbar04**
## Description
 - Add pymc support to predictive_explorer.
 - Add parameter `engine` to predictive_explorer.
 - Add test in test_predictive_explorer.py.
- Update docs

## Checklist

- [x] Code style is correct (follows pylint and black guidelines)
- [x] Includes new or updated tests to cover the new feature
- [x] New features are properly documented (with an example if appropriate)
- [x] Includes a sample plot to visually illustrate the changes (only for plot-related functions)
- [x] Widget states have been properly saved (only for notebooks with widgets) [see](https://github.com/arviz-devs/preliz/wiki/Save-widgets-state) for details.

### CODE REVIEW

File: preliz/internal/plot_helper.py [hunk 1]
```diff
@@ -467,6 +501,50 @@ def plot_repr(results, kind_plot, references, iterations, ax):
...
+    alpha = max(0.01, 1 - iterations * 0.009)
+
+    if kind_plot == "hist":
+        if results[0].dtype.kind == "i":
+            bins = np.arange(np.min(results), np.max(results) + 1.5) - 0.5
+            if len(bins) < 30:
+                ax.set_xticks(bins + 0.5)
+        else:
+            bins = "auto"
+        reshaped_t = np.reshape(results.T, (-1, results.shape[0]))
```
Comment (aloctavodia) [inline preliz/internal/plot_helper.py:514]:
```suggestion

```

File: preliz/internal/plot_helper.py [hunk 2]
```diff
@@ -467,6 +501,50 @@ def plot_repr(results, kind_plot, references, iterations, ax):
...
+        ax.hist(
+            reshaped_t,
```
Comment (aloctavodia) [inline preliz/internal/plot_helper.py:516]:
```suggestion
            results,
```

**Review by aloctavodia — CHANGES_REQUESTED** [review]

**Comment by aloctavodia** [comment]
sample_prior_predictive emits a message. saying "Samplig", plus a list of the sampled variables. We should disable that message. It is generated with the logging module. We should suppress it locally with a context manager, or disable it before sampling and enable it again after sampling. We need to do this to avoid messing up calls to `sample_prior_predictive` outside of predictive_explorer

**Comment by aloctavodia** [comment]
It is no needed for this PR, but to think about it we may want to try to guess the value of "engine", for instance for a pymc model we could inspect the source and detect the regex pattern for "with pm.Model() as" where the particle "pm." is something that can vary.

We may want to keep the "engine" argument and default to "auto" or something similar, just in case our automatic detection fails

File: preliz/internal/plot_helper.py [hunk 3]
```diff
@@ -425,6 +426,39 @@ def looper(*args, **kwargs):
...
+            auto = False
+
+        for _ in range(iterations):
+            with func(*args, **kwargs) as model:
+                obs_name = model.observed_RVs[0].name
+                idata = sample_prior_predictive()
+                val = extract(idata, group="prior_predictive")[obs_name].values
+            if not np.isnan(val).any():
+                results.append(val)
+        results = np.array(results)
```
[the remark this answers is not in the document]
Comment (rohanbabbar04) [reply preliz/internal/plot_helper.py]: Got it 
Thanks...

**Comment by rohanbabbar04** [comment]
> sample_prior_predictive emits a message. saying "Samplig", plus a list of the sampled variables. We should disable that message. It is generated with the logging module. We should suppress it locally with a context manager, or disable it before sampling and enable it again after sampling. We need to do this to avoid messing up calls to `sample_prior_predictive` outside of predictive_explorer

Created a contextmanager to handle the logs generated during this step.

**Comment by rohanbabbar04** [comment]
> It is not needed for this PR, but something to think about.
> 
> We may want to try to guess the value of "engine", for instance for a pymc model we could inspect the source and detect the regex pattern for "with pm.Model() as" where the particle "pm." is something that can vary. For Bambi it should also be simple to detect it.
> 
> We may want to keep the "engine" argument and default to "auto" or something similar, just in case our automatic detection fails

Sounds good.... Let me create a new issue for this

**Comment by rohanbabbar04** [comment]
I am currently keeping the sampling pymc code separated from the preliz code, like it was discussed in the jupyter notebook.
Better to not disrupt the current implementation of the already existing preliz code.

**Comment by aloctavodia** [comment]
agreed, we can refactor and combine later

**Comment by rohanbabbar04** [comment]
> agreed, we can refactor and combine later. Something really easy to combine is the plotting part, right?

Yes...

**Pull request merged**